From 92c4bd208c4c0752bc02d3f11742c3a04aa6f13e Mon Sep 17 00:00:00 2001
Message-ID: <92c4bd208c4c0752bc02d3f11742c3a04aa6f13e.1771966444.git.lorenzo@kernel.org>
In-Reply-To: <2cd15ef3379025ea1ae1f6c36c70621aa578c270.1771966444.git.lorenzo@kernel.org>
References: <2cd15ef3379025ea1ae1f6c36c70621aa578c270.1771966444.git.lorenzo@kernel.org>
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Tue, 24 Feb 2026 17:24:03 +0100
Subject: [bpf-next 03/16] net: airoha: Rely on net_device pointer in ETS
 callbacks

This is a preliminary patch to support multiple net_devices connected
to the same Frame Engine GDM port via an external arbiter.

Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---
 drivers/net/ethernet/airoha/airoha_eth.c | 29 ++++++++++++------------
 1 file changed, 15 insertions(+), 14 deletions(-)

--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -2163,10 +2163,11 @@ airoha_ethtool_get_rmon_stats(struct net
 	} while (u64_stats_fetch_retry(&port->stats.syncp, start));
 }
 
-static int airoha_qdma_set_chan_tx_sched(struct airoha_gdm_port *port,
+static int airoha_qdma_set_chan_tx_sched(struct airoha_gdm_dev *dev,
 					 int channel, enum tx_sched_mode mode,
 					 const u16 *weights, u8 n_weights)
 {
+	struct airoha_gdm_port *port = dev->port;
 	int i;
 
 	for (i = 0; i < AIROHA_NUM_TX_RING; i++)
@@ -2198,16 +2199,16 @@ static int airoha_qdma_set_chan_tx_sched
 	return 0;
 }
 
-static int airoha_qdma_set_tx_prio_sched(struct airoha_gdm_port *port,
+static int airoha_qdma_set_tx_prio_sched(struct airoha_gdm_dev *dev,
 					 int channel)
 {
 	static const u16 w[AIROHA_NUM_QOS_QUEUES] = {};
 
-	return airoha_qdma_set_chan_tx_sched(port, channel, TC_SCH_SP, w,
+	return airoha_qdma_set_chan_tx_sched(dev, channel, TC_SCH_SP, w,
 					     ARRAY_SIZE(w));
 }
 
-static int airoha_qdma_set_tx_ets_sched(struct airoha_gdm_port *port,
+static int airoha_qdma_set_tx_ets_sched(struct airoha_gdm_dev *dev,
 					int channel,
 					struct tc_ets_qopt_offload *opt)
 {
@@ -2249,14 +2250,16 @@ static int airoha_qdma_set_tx_ets_sched(
 	else if (nstrict < AIROHA_NUM_QOS_QUEUES - 1)
 		mode = nstrict + 1;
 
-	return airoha_qdma_set_chan_tx_sched(port, channel, mode, w,
+	return airoha_qdma_set_chan_tx_sched(dev, channel, mode, w,
 					     ARRAY_SIZE(w));
 }
 
-static int airoha_qdma_get_tx_ets_stats(struct airoha_gdm_port *port,
+static int airoha_qdma_get_tx_ets_stats(struct airoha_gdm_dev *dev,
 					int channel,
 					struct tc_ets_qopt_offload *opt)
 {
+	struct airoha_gdm_port *port = dev->port;
+
 	u64 cpu_tx_packets = airoha_qdma_rr(port->qdma,
 					    REG_CNTR_VAL(channel << 1));
 	u64 fwd_tx_packets = airoha_qdma_rr(port->qdma,
@@ -2271,9 +2274,10 @@ static int airoha_qdma_get_tx_ets_stats(
 	return 0;
 }
 
-static int airoha_tc_setup_qdisc_ets(struct airoha_gdm_port *port,
+static int airoha_tc_setup_qdisc_ets(struct net_device *netdev,
 				     struct tc_ets_qopt_offload *opt)
 {
+	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	int channel;
 
 	if (opt->parent == TC_H_ROOT)
@@ -2284,12 +2288,12 @@ static int airoha_tc_setup_qdisc_ets(str
 
 	switch (opt->command) {
 	case TC_ETS_REPLACE:
-		return airoha_qdma_set_tx_ets_sched(port, channel, opt);
+		return airoha_qdma_set_tx_ets_sched(dev, channel, opt);
 	case TC_ETS_DESTROY:
 		/* PRIO is default qdisc scheduler */
-		return airoha_qdma_set_tx_prio_sched(port, channel);
+		return airoha_qdma_set_tx_prio_sched(dev, channel);
 	case TC_ETS_STATS:
-		return airoha_qdma_get_tx_ets_stats(port, channel, opt);
+		return airoha_qdma_get_tx_ets_stats(dev, channel, opt);
 	default:
 		return -EOPNOTSUPP;
 	}
@@ -2840,12 +2844,9 @@ static int airoha_tc_setup_qdisc_htb(str
 static int airoha_dev_tc_setup(struct net_device *netdev,
 			       enum tc_setup_type type, void *type_data)
 {
-	struct airoha_gdm_dev *dev = netdev_priv(netdev);
-	struct airoha_gdm_port *port = dev->port;
-
 	switch (type) {
 	case TC_SETUP_QDISC_ETS:
-		return airoha_tc_setup_qdisc_ets(port, type_data);
+		return airoha_tc_setup_qdisc_ets(netdev, type_data);
 	case TC_SETUP_QDISC_HTB:
 		return airoha_tc_setup_qdisc_htb(netdev, type_data);
 	case TC_SETUP_BLOCK:
